File: backend/src/main/java/com/wolyh/game/backend/dto/Notification.java
package com.wolyh.game.backend.dto;

import java.util.List;
import java.util.Set;

import com.wolyh.game.backend.model.HexCoordinates;
import com.wolyh.game.backend.model.UnitCoordinates;

public record Notification <T> (Type type, T data) {

    public static interface RoomEvent {}
    public static interface GameEvent {}

    public static record GameStart(
        String player1, 
        String player2, 
        Set<String> fov, 
        String roomId
    ) implements RoomEvent {}

    public static record PlayerJoin(
        String username, 
        String roomId
     ) implements RoomEvent {}

    public static record PlayerLeave(
        String username, 
        String roomId
    ) implements RoomEvent {}

    public static record RoomDelete(
        String roomId
    ) implements RoomEvent {}

    public static record AllyUnitMove (
        int unitIdx, 
        List<HexCoordinates> path, 
        List<Set<String>> pathFov,
        List<List<UnitCoordinates>> visibleUnitsAlongPath,
        String roomId
    ) implements GameEvent{}

    public static record EnemyUnitMove(
        int unitIdx, 
        List<HexCoordinates> path, 
        String roomId
    ) implements GameEvent {}

    public static record UnitAttack(
        int attackerIdx, 
        HexCoordinates targetCoords, 
        Set<String> fov, 
        String roomId
    ) implements GameEvent {}

    public static record TurnChange(
        int nextUnitIdx, 
        String roomId
    ) implements GameEvent {}

    public static record MapShrink(
        int shrinkLevel, 
        List<Integer> deadUnits, 
        Set<String> fov, 
        String roomId
    ) implements GameEvent {}

    public static record GameOver(
        String winner, 
        String roomId
    ) implements RoomEvent, GameEvent{}
 
    public static enum Type {
        PLAYER_JOIN,
        PLAYER_LEAVE,
        ROOM_DELETE,
        GAME_START,
        ALLY_MOVE,
        ENEMY_MOVE,
        UNIT_ATTACK,
        TURN_CHANGE,
        MAP_SHRINK,
        GAME_OVER
    }
}

File: frontend/src/dto/Notification.ts
export type Notification = {type: "PLAYER_JOIN", data: PlayerJoinData} |
    {type: "ROOM_DELETE", data: RoomDeleteData} |
    {type: "PLAYER_LEAVE", data: PlayerLeaveData} |
    {type: "ROOM_DELETE", data: RoomDeleteData} |
    {type: "GAME_START", data: GameStartData} |
    {type: "UNIT_ATTACK", data: UnitAttackData} |
    {type: "ALLY_MOVE", data: AllyMoveData} |
    {type: "ENEMY_MOVE", data: EnemyMoveData} |
    {type: "TURN_CHANGE", data: TurnChangeData} |
    {type: "MAP_SHRINK", data: MapShrinkData} |
    {type: "GAME_OVER", data: GameOverData};

export interface PlayerJoinData {
    username: string, 
    roomId: string
}

export interface RoomDeleteData {
    roomId: string
}

export interface PlayerLeaveData {
    username: string, 
    roomId: string
}

export interface GameStartData {
    fov: Array<string>,
    unitSpawns: Array<{idx: number, q: number, r: number}>,
    nb_units: number,
    roomId: string
}

export interface UnitAttackData {
    attackerIdx: number, 
    targetCoords: {q: number; r: number;}, 
    fov: Array<string>, 
    roomId: string
}

export interface AllyMoveData {
    unitIdx: number, 
    path: Array<{q: number; r: number;}>, 
    pathFov: Array<Array<string>>, 
    visibleUnitsAlongPath: Array<Array<{idx: number, q: number, r: number}>>, 
    roomId: string
}

export interface EnemyMoveData {
    unitIdx: number,
    path: Array<{q: number; r: number;}>, 
    roomId: string

}
export interface TurnChangeData {
    nextUnitIdx: number, 
    roomId: string
}

export interface MapShrinkData {
    shrinkLevel: number, 
    deadUnits: Array<number>, 
    fov: Array<string>, 
    roomId: string
}

export interface GameOverData {
    winner: string, 
    roomId: string
}